File: gitdot-web/packages/api/src/endpoint/review/publish-review-comments.ts
import { z } from "zod";
import { ReviewCommentResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ReviewCommentInput = z.object({
  diff_id: z.string().uuid(),
  revision_id: z.string().uuid(),
  body: z.string(),
  file_path: z.string().optional(),
  line_number_start: z.number().int().optional(),
  line_number_end: z.number().int().optional(),
  start_character: z.number().int().optional(),
  end_character: z.number().int().optional(),
  side: z.string().optional(),
});
export type ReviewCommentInput = z.infer<typeof ReviewCommentInput>;

export const PublishReviewCommentsRequest = z.object({
  comments: z.array(ReviewCommentInput),
});
export type PublishReviewCommentsRequest = z.infer<
  typeof PublishReviewCommentsRequest
>;

export const PublishReviewCommentsResponse = z.array(ReviewCommentResource);
export type PublishReviewCommentsResponse = z.infer<
  typeof PublishReviewCommentsResponse
>;

export const PublishReviewComments = {
  path: "/repository/{owner}/{repo}/review/{number}/comments",
  method: "POST",
  request: PublishReviewCommentsRequest,
  response: PublishReviewCommentsResponse,
} as const satisfies Endpoint;
export type PublishReviewComments = typeof PublishReviewComments;

File: gitdot-api/src/endpoint/review/publish_review_comments.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{endpoint::Endpoint, resource::review::ReviewCommentResource};

pub struct PublishReviewComments;

impl Endpoint for PublishReviewComments {
    const PATH: &'static str = "/repository/{owner}/{repo}/review/{number}/comments";
    const METHOD: http::Method = http::Method::POST;

    type Request = PublishReviewCommentsRequest;
    type Response = PublishReviewCommentsResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct ReviewCommentInput {
    pub diff_id: Uuid,
    pub revision_id: Uuid,
    pub body: String,
    pub file_path: Option<String>,
    pub line_number_start: Option<i32>,
    pub line_number_end: Option<i32>,
    pub start_character: Option<i32>,
    pub end_character: Option<i32>,
    pub side: Option<String>,
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct PublishReviewCommentsRequest {
    pub comments: Vec<ReviewCommentInput>,
}

pub type PublishReviewCommentsResponse = Vec<ReviewCommentResource>;
